_posts/2017-11-17-crates-are-not-safe.md

```diff
@@ -0,0 +1,100 @@
+---
+title: Crates.io Ecosystem Not Ready for Embedded Rust
+author: bradjc
+authors: bradjc
+---
+
+As both Tock and the Rust ecosystem grow, we have been looking into using
+community-provided crates on [crates.io](https://crates.io) to allow us to add
+features to Tock. For example, we are exploring using the
+[hmac](https://www.crates.io/crates/hmac) crate to support one of the networking
+protocols we are adding. While this seems promising, we are running into two
+major problems that are preventing us from using crates.io: `unsafe` and
+`std::`.
+
+## Crates are Mostly Unsafe
+
+Tock's kernel is designed to be split into two parts: a minimal but trusted core
+kernel and a much larger set of untrusted "capsules". The core kernel handles
+loading applications and controlling memory-mapped IO registers, and as such
+requires using `unsafe` code. The capsules, however, implement chip drivers and
+other utilities, and are strictly bound by the Rust typesystem (i.e. no
+`unsafe`). This divide allows us to maximize the safety guarantees of Rust while
+still making a kernel possible.
+
+This restriction on capsules has so far made it very difficult to use
+community-developed crates. Following up on @josh_triplett's [analysis of
+crates.io](https://twitter.com/josh_triplett/status/849089108727222275), we
+survey all of the crates on crates.io and find that 34% (4216/12360) of the
+crates use an `unsafe` block at least once. Of crates that have been downloaded
+10k times or more, 55% (522/944) use `unsafe`. While these ratios are not
+promising, it may be possible to work around the unsafe crates. The problem gets
+worse, however, when we consider dependencies. A crate that does not use
+`unsafe` but relies on a crate that does is effectively unsafe itself. After
```

Comment:
"Even if a crate, itself, doesn't use `unsafe`, if one of it's dependencies does, we still can't use it."